### DISCUSSION TITLE
Issue troubleshooting an s3 connection for remote logging to Min.io

### DISCUSSION THREAD

**Discussion by hanleybrand** [apache/airflow#26979] [Q&A] [answered]
I'm trying to configure a Kubernetes deployment of Airflow 2.4.1 to use remote logging, but I can't create a working Amazon Web Services connection in the Airflow UI (or at least I get an error when I test the connection)
 
### TLDR; 

While searching and hacking around seem to have solved some of the issue (or at least fixed some other issues I hadn't been aware of?) when I use either the UI or python route to test_connection I get an error, but every other way I connect seems to work. 

The error I get when testing the connection is: 

> **An error occurred (InvalidParameterValue) when calling the GetCallerIdentity operation: Unsupported action GetCallerIdentity** 

The biggest issue of course, is that when I deploy/redeploy to the kubernetes cluster logs are not being written to the 3s bucket, which makes it difficult to check the logs to see what's going on 😉 

My guess at the problem when I look at the code (  [AwsGenericHook](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/base_aws.py#L631)?  ) that I think is generating that error, the test_connection function is checking the connection using STS (session tags), which I don't have configured (I don't know if min.io supports STS actually, it looks like it supports AssumeRoleWithWebIdentity )

### My configuration:

Anonymizing my config particulars, let's say Airflow is deployed to `https://airflow.k8s.example.com`

In Admin -> Connections  I have a conn_id 'amazon_s3' of type 'Amazon Web Services' with the correct AWS Access Key ID & Secret Access Keys set, plus Extra set to: 

```
{ "aws_access_key_id": "LAWGIN", 
  "aws_secret_access_key": "PASSWOID", 
  "endpoint_url": "https://minio01.k8s.example.com:9000", 
   "region_name": "us-east-1"
   "encrypted": "yes", "verify": "False"}
```

**Aiflow.cfg logging section** 

```cfg
[logging]
colored_console_log = False
encrypt_s3_logs = False
remote_base_log_folder = s3://airflow-logs-dev
remote_log_conn_id = amazon_s3
remote_logging = True
```

**Notes**:  

- "encrypted" set to "no" or "yes" does not change what error I got.  
- "verify" can likewise be set to "False", "None" or "True" -- the only thing that changes the error is removing the key/value pair for verify  
- I can also remove the login/password k/vs from the Exrta JSON string, it doesn't seem to matter. I leave them in because some of the discussions implied that they needed to be in the extra string.

With "verify" key included I get the following error:

> **- SSL validation failed for https://minio01.k8s.example.com:9000/ [Errno 2] No such file or directory**

If I remove the  "verify" key entirely, I also get this error.

> **An error occurred (InvalidParameterValue) when calling the GetCallerIdentity operation: Unsupported action GetCallerIdentity** 

### HIstory, stuff I've tried, etc

I had started with what the docs say, which was the simpler extra set to 
{ "aws_access_key_id": "LAWGIN",  "aws_secret_access_key": "PASSWOID", 
 "host": "https://minio01.k8s.example.com:9000"} but I got errors regarding the region not being set correctly.

I have confirmed with the aws-cli and min.io mc clients that the aws_access_key_id, aws_secret_access_key and endpoint_url paremters above are correct, e.g.:

 ```shell
aws --endpoint-url https://minio01.k8s.example.com:9000 s3 ls s3://airflow-logs-dev
```
and
``` shell
mc alias set aa https://minio01.k8s.example.com:9000 LAWGIN PASSWOID
mc ls aa/airflow-logs-dev
```
both allow me to interact with the Min.io instance, so I am certain the issue is not a typo in my login/password, etc. (I copy/pasted values out of my config to test with aws/mc)

Additionally, I can execute a shell in the webserver pod and using python in the shell:

```python
>>> import boto3
>>> client = boto3.client(
...     's3',
...     aws_access_key_id='LAWGIN',
...     aws_secret_access_key='PASSWOID',
...     region_name='us-east-1',
...     endpoint_url='https://minio01.k8s.example.com:9000',
... )
>>> 
>>> client.list_buckets()
{'ResponseMetadata': {'RequestId': '171CD39BF1DA4102', 'HostId': '', 'HTTPStatusCode': 200, 'HTTPHeaders': {'accept-ranges': 'bytes', 'content-length': '461', 'content-security-policy': 'block-all-mixed-content', 'content-type': 'application/xml', 'server': 'MinIO', 'strict-transport-security': 'max-age=31536000; includeSubDomains', 'vary': 'Origin, Accept-Encoding', 'x-amz-bucket-region': 'us-east-1', 'x-amz-request-id': '171CD39BF1DA4102', 'x-content-type-options': 'nosniff', 'x-xss-protection': '1; mode=block', 'date': 'Mon, 10 Oct 2022 21:50:28 GMT'}, 'RetryAttempts': 0}, 'Buckets': [{'Name': 'airflow-logs-dev', 'CreationDate': datetime.datetime(2022, 10, 3, 16, 26, 7, 938000, tzinfo=tzlocal())},....#etc

```

There was a clue that it it might be certificate related so I tried adding `RUN cp $(python -m certifi) /etc/ssl/certs/` to my Dockerfile, and while it seems to help (no longer get a cert error with `S3Hook.check_for_bucket()`) I'm back to getting the InvalidParameterValue/GetCallerIdentity error 

```python
>>> from airflow.providers.amazon.aws.hooks.s3 import S3Hook
>>> s3h = S3Hook('amazon_s3')
>>> b = 'airflow-logs-dev'
>>> s3h.check_for_bucket(b)
[2022-10-11T02:39:41.134+0000] {base.py:71} INFO - Using connection ID 'amazon_s3' for task execution.
[2022-10-11T02:39:41.137+0000] {connection_wrapper.py:303} INFO - AWS Connection (conn_id='amazon_s3', conn_type='aws') credentials retrieved from login and password.
True
# that's good
>>> s3h.test_connection()
(False, 'An error occurred (InvalidParameterValue) when calling the GetCallerIdentity operation: Unsupported action GetCallerIdentity')
# not as good
```

Is anyone familiar with setting this kind of connection/logging config up on the more recent versions of Airflow?   As far as I can tell the  error is coming from the [AwsGenericHook](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/base_aws.py#L631) class, but it seems like an error with session tokens (STS) which isn't what I'm using in my config, so it's confusing.

edit: edited for clarity, hopefully

**Comment by Taragolis** [comment]
Hey @hanleybrand,
 
Amazon Web Services Connection use for almost all of amazon provider components (operators, hooks, sensors, etc.) except Redshift Connection, so we only could validate credentials by calling [GetCallerIdentity](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html) which always should available if you use AWS with correct credentials. We do not test (yet?) any Identity-based policies and resource-based policies as well as access to specific resource .

I'm not familiar with minio but if it not support STS method GetCallerIdentity then run `test_connection` method doesn't make sense.

**Reply by Taragolis** [reply]
> * "encrypted" set to "no" or "yes" does not change what error I got.
> * "verify" can likewise be set to "False", "None" or "True" -- the only thing that changes the error is removing the key/value pair for verify
> * I can also remove the login/password k/vs from the Exrta JSON string, it doesn't seem to matter. I leave them in because some of the discussions implied that they needed to be in the extra string.

https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html#configuring-the-connection

1. `encrypted` has no affect because is is not exists in `boto3.session.Session` as well as not exists in `boto3.client` and `boto3 resource`
2. Based on [boto3.client doc](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client) `verify` has no affect if `endpoint_url` set.
3. If explicit credentials required then you could provide as login/password or in extra.

**Comment by hanleybrand** [comment]
Hey @Taragolis, thanks for taking the time to answer -- your explanations were really helpful. 

So, the base issue that I've discovered is that min.io apparently hasn't implemented the GetCallerIdentity action, which seems odd -- using an awscli configuration that matches the above, 

```
$ aws --endpoint-url hhttps://minio01.k8s.example.com:9000 sts get-caller-identity

An error occurred (InvalidParameterValue) when calling the GetCallerIdentity operation: Unsupported action GetCallerIdentity
```

😞 Not sure if this means I need to find a different logging solution for my airflow instance, however as I was searching for min.io STS support docs, I had found this doc on [min.io's implementation of the Assume Role API ](https://github.com/minio/minio/blob/master/docs/sts/assume-role.md#using-assumerole-api)which led me to try this: 

```
$ aws  --endpoint-url https://minio01.k8s.example.com:9000 \
    sts assume-role \
  --policy '{"Version":"2012-10-17","Statement":[{"Sid":"Stmt1","Effect":"Allow","Action":"s3:*","Resource":"arn:aws:s3:::*"}]}' \
  --role-arn arn:xxx:xxx:xxx:xxxx \
  --role-session-name airflow

# which returned this, as described in the document
{
    "Credentials": {
        "AccessKeyId": "RANDOMNUMBERSANDLETTERSKEYID",
        "SecretAccessKey": "RANDOMNUMBERSANDLETTERSSECRETKEY",
        "SessionToken": "REALLYCONVINCINGTOKEN",
        "Expiration": "2022-10-12T20:32:30+00:00"
    },
    "AssumedRoleUser": {
        "Arn": ""
    }
}
```

Because of this (and after reading some of the boto3 docs, I modified my Airflow Connection Extra string like so (I was in denial about the GetCallerIdentity method not being implemented I guess):

```
 { 
  "aws_access_key_id": "LAWGIN", "aws_secret_access_key": "PASSWOID",
  "endpoint_url": "https://minio01.k8s.example.com:9000", 
  "region_name": "us-east-1", 
  "role_arn": "arn:xxx:xxx:xxx:xxxx", 
  "policy": {"Version": "2012-10-17", "Statement": [{"Sid": "Stmt1", "Effect": "Allow", "Action": "s3:*", "Resource": "arn:aws:s3:::*"}]}, 
  "role_session_name": "airflow", 
"assume_role_method" : "assume_role"
  }
```

and to my surprise the error changed to 

>  **An error occurred (InvalidClientTokenId) when calling the AssumeRole operation: The security token included in the request is invalid.**

... even though that doesn't make a whole lot of sense at first glance.

But I think what's happening is that within AwsGenericHook.test_connection(),  is evaluating ```res = self.get_client_type().get_caller_identity()```,  self.get_client_type() creates a session, but because the extra inits the AwsGenericHook object with a role_arn attribute already, the BaseSessionFactory.create_session() is returning self._create_session_with_assume_role().... which I would thing would work, although it's not obviously. 

Is this a case where the answer is I should try to create a Minio provider based on the amazon provider, or is it possible there's an additional key/value I can supply via the Extra string that would lead the boto3 session to be created correctly in the context of the AwsGenericHook?

**Comment by Taragolis** [comment, marked as answer]
> Not sure if this means I need to find a different logging solution for my airflow instance

Do not read entire message yet but to clarify main purpose of `test_connection` method in hook only for allow testing connection within the UI/API. All other components provided by `apache-airflow-providers-amazon` (include logging) do not use this method.

**Reply by hanleybrand** [reply]
Does that mean that if `test_connection` is failing, it doesn't necessarily mean that the connection won't work?

**Reply by Taragolis** [reply]
STS service [GetCallerIdentity](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts.html#STS.Client.get_caller_identity) returns only information about user.

```shell
# Wrong credentials (usual aws_access_key_id)
❯ aws sts get-caller-identity

An error occurred (InvalidClientTokenId) when calling the GetCallerIdentity operation: The security token included in the request is invalid.

# Wrong credentials another error (usual aws_secret_access_key or aws_session_token)
❯ aws sts get-caller-identity

An error occurred (SignatureDoesNotMatch) when calling the GetCallerIdentity operation: The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details

# Credentials not set
❯ aws sts get-caller-identity

Unable to locate credentials. You can configure credentials by running "aws configure".

# Everything fine with credentials and it can be use
❯ aws sts get-caller-identity

{
    'Account': '123456789012',
    'Arn': 'arn:aws:iam::123456789012:user/Alice',
    'UserId': 'AKIAI44QH8DHBEXAMPLE'
}
```

---

STS method [AssumeRole](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts.html#STS.Client.assume_role) use for assume another role. Did you internally use this functionality in minio? If not you shouldn't specify `role_arn` in connection.

---

Conclusion. Even if you unable to "test" connection (like in case minio doesn't not support STS GetCallerIdentity) doesn't mean that you can not use s3 compatible logging. You could check in [airflow/providers/amazon/aws/log/s3_task_handler.py](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/log/s3_task_handler.py) which methods of [S3Hook](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/s3.py) uses.

**Reply by Taragolis** [reply]
> Does that mean that if test_connection is failing, it doesn't necessarily mean that the connection won't work?

Yep in your case it is fine. It only make sense in "native" AWS environment because it should be available if credentials correct.

**Reply by Taragolis** [reply]
That also might be a good point to add this in `test_connection` response message.
If I have a time to reproduce this error I will try to add this

**Comment by hanleybrand** [comment]
@Taragolis -- thanks again for all your help, I got it working!

For anyone having trouble and coming across this discussion, here's a python script to run from one of the airflow pods (I used scheduler, but webserver or triggerer work as well) 

using `airflow info` and `env | grep [term]` are useful for checking that settings are correct on your pods 

In case it's not clear, the comments under the lines are output you should see if the connection is set up correctly so that task logging will happen as expected

For reference, the Airflow Connection is called `minio_s3`, is type `Amazon AWS`, and only has the extra field set, like so:

```json
{ 
  "aws_access_key_id": "LAWGIN", 
   "aws_secret_access_key": "PASSWOID",
  "endpoint_url": "https://minio01.k8s.example.com:9000", 
  "region_name": "us-east-1"
}
```

Here's the test script (obvs edit the connection name and bucket name to fit your config)

```python
### from $ airflow info:
# Apache Airflow
# task_logging_handler   | airflow.providers.amazon.aws.log.s3_task_handler.S3TaskHandler
# remote_base_log_folder | s3://airflow-logs-dev/logs
# Providers info
# apache-airflow-providers-amazon          | 6.0.0

### ENV VARS ( $ env | grep LOG )
#   AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER=s3://airflow-logs-dev/logs
#   AIRFLOW__LOGGING__REMOTE_LOG_CONN_ID=amazon_s3
#   AIRFLOW__LOGGING__REMOTE_LOGGING=True

### prior to running this script, use minio's mc to stick a test file in the minio bucket
#   $ mc alias set minio_s3 https://minio01.k8s.example.com:9000 LAWGIN PASSWOID
#   $ echo "hello" > hello.txt
#   $ mc cp hello.txt minio_s3/airflow-logs-dev/

from airflow.providers.amazon.aws.hooks.s3 import S3Hook

cid = "minio_s3"                       # the airflow connection id
b_url = 's3://airflow-logs-dev/logs'   # remote_base_log_folder

hook = S3Hook(cid, transfer_config_args={
    'use_threads': False,
})

parsed_loc = hook.parse_s3_url(b_url)[0]
# 'airflow-logs-dev'
b, k = hook.parse_s3_url(b_url)
# b == 'airflow-logs-dev', k == 'logs'
hook.get_s3_bucket_key('airflow-logs-dev', 'logs', b, k)
# ('airflow-logs-dev', 'logs')

hook.check_for_bucket(parsed_loc)  # parsed_loc == 'airflow-logs-dev'
# [2022-10-20T16:28:15.808+0000] {base.py:71} INFO - Using connection ID 'amazon_s3' for task execution.
# [2022-10-20T16:28:15.868+0000] {connection_wrapper.py:306} INFO -
#       AWS Connection (conn_id='amazon_s3', conn_type='aws') credentials retrieved from extra.
# True
hook.check_for_bucket(b)
# True
hook.check_for_bucket(k)
# [2022-10-20T16:31:37.605+0000] {s3.py:223} ERROR - Bucket "logs" does not exist
# False
# Expected - k is a key, not a bucket

print(hook.list_keys(b))
# ['hello.txt'] -- only if you uploaded the file like in the comment above and that's all that's in there
```

**Comment by cheeyeelim** [comment]
Other S3-compatible storage like Digital Ocean Spaces works with a similar configuration as well. Below is the configuration I used, all set as environment variables (for ease of k8s deployment).

```
AIRFLOW__LOGGING__REMOTE_LOGGING=True
AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER='s3://<BUCKET_NAME>/logs'
AIRFLOW__LOGGING__REMOTE_LOG_CONN_ID='s3_do_conn'
AIRFLOW_CONN_S3_DO_CONN='{
  "conn_type": "aws",
  "login": "<ACCESS_KEY>",
  "password": "<SECRET_ACCESS_KEY>",
  "extra": {
    "region_name": "<REGION_NAME>",
    "endpoint_url": "https://<REGION_NAME>.digitaloceanspaces.com"
  }
}'
```

It took me a few hours to figure this out. Similarly I was stumped by constant testing errors from `airflow connections test`. Getting errors from the test does not mean that the connection will not work for remote logging.

**Reply by fayadchowdhury** [reply]
This is exactly what I needed! Thanks a bunch!
